src/mobu/storage/git.py

```diff
@@ -0,0 +1,255 @@
+    config_location
+        Filesystem path for the file to use as the user-global Git config.
+    logger
+        Logger to use.
+    """
+
+    def __init__(
+        self,
+        *,
+        repo: Path | None = None,
+        config_location: Path | None = None,
+        logger: BoundLogger | None = None,
+    ) -> None:
+        self.repo = repo
+        self._logger = logger
+        self._config_location = config_location
+
+    async def _exec(
+        self,
+        cmd: str,
+        *args: str,
+        env: dict[str, str] | None = None,
+        cwd: Path | None = None,
+    ) -> None:
+        """Execute a non-interactive subprocess.
+
+        We need this to make the git commands behave correctly in an
+        async environment.  The reason why is a little subtle.  Most git
+        commands let you specify arbitrary paths, but not `git add`.  To
+        run a git add, you have to have a working tree, and that means that
+        the current working directory has to be inside that working tree.
+
+        The current working directory is a per-process global.  So if we
+        didn't do this, then mobu, which may be running many tests at once
+        in different coroutines, will get very confused.
+
+        However, since we're creating a subprocess to run the git
+        executable anyway, we can run it with asyncio.subprocess,
+        optionally with the working directory set to an arbitrary
+        path.  That way, mobu's working directory stays untouched, but
+        the subprocess may execute somewhere else.
+        """
+        l_args = [cmd]
+        l_args.extend(args)
+        cmd_and_args = join(l_args)
+        sub_env = os.environ.copy()
+        if env is not None:
+            sub_env.update(env)
+
+        proc = await asyncio.subprocess.create_subprocess_exec(
+            cmd,
+            *args,
+            cwd=cwd,
+            stdin=asyncio.subprocess.DEVNULL,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.PIPE,
+            env=sub_env,
+        )
+        stdout, stderr = await proc.communicate()  # Waits for process exit
+
+        stdout_text = ""
+        stderr_text = ""
+        if proc.stdout is not None:
+            stdout_text = stdout.decode()
+        if proc.stderr is not None:
+            stderr_text = stderr.decode()
+        if proc.returncode != 0:
+            raise SubprocessError(
+                f"Subprocess '{cmd_and_args}' failed",
+                returncode=proc.returncode,
+                stdout=stdout_text,
+                stderr=stderr_text,
+                cwd=cwd,
+                env=sub_env,
+            )
+        if self._logger:
+            msg = (
+                f"'{cmd_and_args}' exited with rc={proc.returncode};"
+                f" stdout='{stdout_text}'; stderr='{stderr_text}'"
+                f" working_directory='{cwd!s}'; env='{sub_env}'"
+            )
+            self._logger.debug(msg)
```

Comment:
All those key/value pairs should be key/value arguments to `debug`, not collapsed into a string. We're using structlog exactly so that we can preserve structure like this as JSON in the log message.